### PR TITLE
Integration Testing Framework: Add retries for artifacts API call

### LINKED ISSUE #4345: Integration Testing Framework: Add retries to Artifacts API call
Currently in the integration testing framework, the `ArtifactsAPIClient` doesn't retry calls to the Artifacts API. 
 Speaking with @davesys911, he suggested we should add retries to this call, somewhere in this function:

https://github.com/elastic/elastic-agent/blob/f9310c7f664bb25f37d671cca4e2fcf3749732c8/pkg/testing/tools/artifacts_api.go#L306-L319

### PR DESCRIPTION
**Pull request by ycombinator**
## What does this PR do?

This PR adds retries to the Artifacts API call in the integration testing framework, as recommended by @DaveSys911.

## Why is it important?

The Artifacts API can sometimes be flaky. 

## Related issues
- Resolves #4345

### CODE REVIEW

**Comment by ycombinator** [comment]
Moving PR into draft to a) update unit test assertions and b) refactor logging logic so logger doesn't need to be passed down in several places.

File: pkg/testing/tools/artifacts_api.go [hunk 1]
```diff
@@ -310,9 +317,27 @@ func (aac ArtifactAPIClient) createAndPerformRequest(ctx context.Context, URL st
 		return nil, err
 	}
 
-	resp, err := aac.c.Do(req)
+	// Make the request with retries as the artifacts API can sometimes be flaky.
+	var resp *http.Response
+	// TODO (once we're on Go 1.22): replace with for numAttempts := range maxAttemptsForArtifactsAPICall {
+	for numAttempts := 0; numAttempts < maxAttemptsForArtifactsAPICall; numAttempts++ {
+		resp, err = aac.c.Do(req)
+		if err == nil {
```
Comment (cmacknz) [inline pkg/testing/tools/artifacts_api.go:325]: You should not continue to retry if the error is context.Cancelled or context.DeadlineExceeded.

Comment (ycombinator) [reply pkg/testing/tools/artifacts_api.go:327]: Fixed in adf118f6cd.

**Review by cmacknz — APPROVED** [review]

**Review by rdner — COMMENTED** [review]
@ycombinator Unfortunately, I didn't look at this PR in time. Sorry about that.

I don't think this is how we should approach it, we should:

1. Implement the `httpDoer` interface with retries and use this implementation as an option
2. Continue using the existing option pattern and remove the logger parameter.
3. The logger interface is not implemented by any logger, only by `testing.T` and this client is used beyond testing.

I addressed 2 and 3 in my PR https://github.com/elastic/elastic-agent/pull/4331 (had to resolve some conflicts)

For 1 @pchila filed an issue here https://github.com/elastic/elastic-agent/issues/4354

**Pull request merged**